frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java

```diff
@@ -46,185 +45,217 @@
         }
       }
+    }
+  }
+
+  private void filterNodesByRackSelectionPolicy(ResourceDescriptor requestResource, int pendingTaskNumber) {
+    //TODO: Node GPU policy filter the nodes;
+  }
 
-      // Found a selectionResult passed all the Tests above
-      selectionResult = new SelectionResult();
-      selectionResult.setNodeHost(nodeHost);
-      selectionResult.setGpuAttribute(optimizedRequestResource.getGpuAttribute());
-      break;
+  private SelectionResult SelectNodes(ResourceDescriptor requestResource, int pendingTaskNumber) {
+    //TODO: apply other node selection policy in  the futher;
+    return selectNodesByJobPacking(requestResource, pendingTaskNumber);
+  }
+
+  //Default Node Selection strategy.
+  private SelectionResult selectNodesByJobPacking(ResourceDescriptor requestResource, int pendingTaskNumber) {
+
+    int requestNumber = pendingTaskNumber * am.getConfiguration().getLauncherConfig().getAmSearchNodeBufferFactor();
+    List<Node> candidateNodes = new ArrayList<Node>();
+    SelectionResult result = new SelectionResult();
+    for (String nodeName : filteredNodes) {
+      candidateNodes.add(allNodes.get(nodeName));
+    }
+    Collections.sort(candidateNodes);
+    for (int i = 0; i < requestNumber && i < candidateNodes.size(); i++) {
+      Node select = candidateNodes.get(i);
+      Long gpuAttribute = requestResource.getGpuAttribute();
+      if (gpuAttribute == 0) {
+        gpuAttribute = selectCandidateGpuAttribute(select, requestResource.getGpuNumber());
+      }
+      result.addSelection(select.getHost(), gpuAttribute, select.getAvailableResource().getPortRanges());
     }
+    return result;
+  }
 
-    if (selectionResult != null) {
-      LOGGER.logInfo(
-          "select: Found a SelectionResult satisfies the Request: SelectionResult: [%s]",
-          selectionResult);
-    } else {
-      LOGGER.logWarning(
-          "select: Cannot found a SelectionResult satisfies the Request: " +
-              "Check whether the Request can be relaxed to RM");
-      String notRelaxLogPrefix = "select: The Request cannot be relaxed to RM: Reason: ";
-
-      // Test NodeGpuType
-      if (requestNodeGpuType != null) {
-        throw new NotAvailableException(
-            String.format(notRelaxLogPrefix +
-                    "NodeGpuType is specified: Request NodeGpuType: [%s]",
-                requestNodeGpuType));
+  public synchronized SelectionResult select(ResourceDescriptor requestResource, String taskRoleName) throws NotAvailableException {
+
+    LOGGER.logInfo(
+        "select: TaskRole: [%s] Resource: [%s]", taskRoleName, requestResource);
+    String requestNodeLabel = am.getRequestManager().getTaskPlatParams().get(taskRoleName).getTaskNodeLabel();
+    String requestNodeGpuType = am.getRequestManager().getTaskPlatParams().get(taskRoleName).getTaskNodeGpuType();
+    int pendingTaskNumber = am.getStatusManager().getUnAssociatedTaskCount(taskRoleName);
+    List<ValueRange> reUsePorts = null;
```

Comment:
>reUsePorts [](start = 21, length = 10)

reusePorts

Reply:
This not only means the assiciatedPorts it is also means the requestedPorts

---
In reply to: [174711836](https://github.com/Microsoft/pai/pull/318#discussion_r174711836) [](ancestors = 174711836)

Reply:
reusePorts

---
In reply to: [175340387](https://github.com/Microsoft/pai/pull/318#discussion_r175340387) [](ancestors = 175340387,174711836)